### DISCUSSION TITLE
Slober compile messages:  LinkNameProvder and ArchiveNameProvider

### DISCUSSION THREAD

**Discussion by pcarew** [Sloeber/arduino-eclipse-plugin#1640] [Q&A] [answered]
What does the error messages:
"**LinkNameProvider: The call should go to the overloaded function not here**" and 
"**ArchiveNameProvider: The call should go to the overloaded function not here.**"

refer two?
I'm getting them, but it isn't stopping a successful compile.

I have a multi class file project for the Arduino. I have developed a RealTime OS that I am using. This necessitates that two of the CPP files have a different optimization level. For these two files, I have added -O0 to the end of the command line in the file properties (C/C++Build->Settings->Tool Settings->CommandLinePattern).
Now when I compile I get the two messages above, but the compile is completed successfuly.

Do I just ignore these messages or can someone please explain what they're refering to?
Thank you
Paul.

**Comment by jantje** [comment, marked as answer]
Some hard to answer questions.
If you look at the LinkNameProvider code you will see it is derived from IManagedOutputNameProviderJaba (jaba is me so this is Sloeber code)
https://github.com/Sloeber/arduino-eclipse-plugin/blob/1c5136eaaf1a448bd1fbea2c69f42f99baa79afe/io.sloeber.core/src/io/sloeber/core/toolchain/LinkNameProvider.java#L19
IManagedOutputNameProviderJaba  is derived from IManagedOutputNameProvider (this is managed build code aka not from Sloeber)  and adds a getOutputNames with more parameters.
https://github.com/Sloeber/arduino-eclipse-plugin/blob/1c5136eaaf1a448bd1fbea2c69f42f99baa79afe/io.sloeber.core/src/io/sloeber/core/toolchain/IManagedOutputNameProviderJaba.java
Basically Sloeber needed more information to make the correct decision on the nameProviders so sloeber extended the managedBuild API . However managed Build does not know nor care about this.
The standard CDT managed build make file generator uses therefore uses the simple getOutputNames but the Sloeber makefile generator uses the complex getOutputNames. 
As LinkNameProvider should only be used in sloeber projects using the sloeber makefile generator it looks like in your setup there is a managed build (read not Sloeber) call to the Sloeber LinkNameProvider which should not happen.

Which bring me to
>For these two files, I have added -O0 to the end of the command line in the file properties (C/C++Build->Settings->Tool Settings->CommandLinePattern).

I never ever used/tested this (congrats for thinking about this) and it may be the root cause of the error messages popping up. To me it seems likely that this activates some managedBuild code that normally is not activated and that causes the errors.
How is that possible? Sloeber makefile generator is a copy of the managed Build makefile generator and sloeber only uses a small subset of the code but all the code is still there.

>Do I just ignore these messages or can someone please explain what they're refering to?

I hope the above explanation explains what the error messages are referring to.
You can ignore them if the actual build commands executed are correct for you.

**Reply by pcarew** [reply]
Thank you Jantje for a thoughtful and comprehensive reply.

It definitely **is** the addition of the "-O0" at the end of the command line in the file properties build settings entry. If I take this -O0 back off, the error disappears. It seems that I can ignore the messages, but they do appear in a pop-up which is mildly irritating to to keep clearing/dismissing

Is there a better place to add 'per file' GCC command line option changes/additions that maybe wouldn't trigger these messages?
Whereabouts are the command line FLAG macros defined?
I thought about seeing if I could add or amend those FLAG macro/variable definitions, but I wasn't sure where they get defined.

Thank you for a stable and **highly useful** addition to Eclipse for Adruino development.
Paul.

**Reply by jantje** [reply]
>Is there a better place to add 'per file' GCC command line option changes/additions

The thing is ... Sloeber does not support per file settings. What you do is a unsupported workaround.

> but they do appear in a pop-up which is mildly irritating to to keep clearing/dismissing

Best way forward it to build sloeber yourself and remove the error generations.
No need to fix the real issue as I'm currently completely rewriting this part.

**Reply by pcarew** [reply]
Ok, thanks for helping me understand
Paul
